### ISSUE TITLE
Not working with no-neutral light switch

### ISSUE THREAD

**Issue by sarnex** [juanboro/cync2mqtt#18]
Hi, 

I have a no neural light switch from CE. The link is [here](https://www.gelighting.com/led-lights/bulbs/ge-cync-smart-switch-no-neutral-wire-required-paddle-style-bluetooth-and-24-ghz)

I was able to dump the config from GE's API fine.

 When I run the script, I always get:
```
 cync2mqtt - INFO - Unable to connect to mesh mac for notify: 78:6D:EB:F0:5D:C9 - [org.bluez.Error.Failed] Operation failed with ATT error: 0x0e (Unlikely Error)
```

My guess is the reverse engineered BT stuff isn't working for this switch.

Is there any way I can help get this working? I'm relatively technically able myself but I don't know how to reverse engineer bluetooth. I can provide BT packet dumps or any other information if you have time to work on this and can tell me what to do at a high level.

Thanks for this project.

**Comment by juanboro** [comment]
I guess the question is what chipset the switch uses... Some thoughts...
The older cync stuff uses Telink's mesh spec where they developed their own legacy mesh protocol - which the dimond library (https://github.com/google/python-dimond) was written on.   I basically leveraged that + Telink's documentation here: http://wiki.telink-semi.cn/wiki/protocols/Telink-Mesh/ .  I learned a bunch about BLE by looking at the bleak library examples here: https://github.com/hbldh/bleak/tree/develop/examples .  Telink's older mesh helped then influence the development of the standard BLE mesh systems that exist today.

I think there are three possibilities here:
1.  The switch uses the old Telink mesh protocol - but something strange is happening.  To further debug this you would want to further break up the steps in the try/except block in the connect method of mesh.py.  Basically, it's trying to subscribe to notifications from the mesh network.  What is weird about the above error is it _seems_ like it is successfully pairing to the mesh device but it isn't able to subscribe to the notifications about when states change.  So it seems like it is really close to working.
2. It's possible this uses some completely new mesh protocol .... i'd hope that isn't the case.
3. These might use direct-connect.  I've been trying to get time to play with these... but my suspicion is that direct-connect bulbs use the real bluetooth mesh protocol (lots of examples and documentation here: https://www.bluetooth.com/specifications/specs/).  There is a project here: https://github.com/dominikberse/homeassistant-bluetooth-mesh --- which may work with this.  Would be great to see if anyone has had any luck with that.  It's on my to-do list to play with :-).

My experience with debugging BT packet dumps isn't good (at least from BT-mesh) --- I've had more luck with working the  mesh specs/api directly (basically the packets dumps will just follow the processes documented in the specs --- but the contents have to be decrypted according to said specs to know what actually is going on).

**Comment by sarnex** [comment]
Thanks for the quick response.

1. Just for some info, I'm running this on a raspi 3b+. I also tried this on my desktop and got the same exact error, so it doesn't seem to be something with the bluetooth chipset/driver on the raspi which I was hoping for as an easy solution. The line that fails is
```
await self.client.start_notify(atelink_mesh.notification_char, self.callback_handler)
```
I'm not really sure what to do to get more info debugging this, I expect the implementation of the Telink spec in this repo is fine as other devices work.
2. I also hope not
3. So I tried that repo and I wasn't sure exactly how the config file from here maps to the config file from there. I tried some combinations of stuff and I always got 
```
INFO:root:Scanning for unprovisioned devices...

Found 0 nodes:
```
I also tried the above when the device was in pairing state (hold down button starts blinking slowly) and same thing.
If you think that error could be related to a bad config file (they need hass_device_id/bluetooth_mesh_device_uuid/hass_device_name), if you could help me with finding those from my config file for this repo or the cync_data repo dumps that would be helpful.
I also tried a few phone apps that claimed to be able to scan for both provisioned and unprovisioed BT Mesh device and none of them found anything, so maybe it's not BT Mesh. I didn't try the phone apps when in a pairing state though.

Any ideas on what I can do next? 
Thanks again

**Comment by juanboro** [comment]
Yeah - for number 3 above... if an app like nRFMesh didn't see it when it was in pairing (unprovisioned) mode, it likely isn't BT mesh.  I'll try to get sometime this evening to refresh my BT debugging knowledge for a few things to possibly try.

**Comment by sarnex** [comment]
Yep, just confirmed nRFMesh can't see it in pairing mode as either an unprovisioned or provisioned device.

Thanks a lot, I really appreciate it!

**Comment by juanboro** [comment]
ok - couple things... first question - when you get_cync_config_from_cloud -- what does the yaml look like?  this is basically what i see:
```yaml
meshconfig:
  ddddddddd:
    access_key: dddddd
    bulbs:
      1:
        is_plug: true
        mac: AA:BB:CC:DD:EE:01
        name: plug 1
      2:
        mac: AA:BB:CC:DD:EE:02
        name: light 2
    mac: DdddDdddddd
    name: Home
mqtt_url: mqtt://127.0.0.1:1883/
```

Second --- if you do this command:
```shell
 sudo hcitool lescan
```

You should be able to see the telink meshnodes and then the friendly name be the mesh mac, i.e:
```
AA:BB:CC:DD:EE:01  DdddDdddddd
AA:BB:CC:DD:EE:02  DdddDdddddd
```

you should hopefully be able to query one of the nodes: (this only sometimes works for me):
```shell
sudo hcitool leinfo AA:BB:CC:DD:EE:01
```
When it does, I see something like:
```
Requesting information ...
        Handle: 65 (0x0041)
        LMP Version: 4.2 (0x8) LMP Subversion: 0x8
        Manufacturer: Telink Semiconductor Co. Ltd (529)
        Features: 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00
```
I think there is a way to ask it for its gatt characteristics --- will have to play more with bluetoothctl (see: https://stackoverflow.com/questions/47825128/how-to-read-characteristic-from-bluetooth)

**Comment by sarnex** [comment]
Thanks for the response.

Here's my config, with all info changed obviously.
```
meshconfig:
  1111111111:
    access_key: 222222
    bulbs:
      1:
        mac: AA:AA:AA:AA:AA:AA
        name: Switch
    mac: 3333333333333333
    name: why
mqtt_url: mqtt://username:pass@127.0.0.1:1883/
```

lescan output:
```
...
AA:AA:AA:AA:AA:AA 3333333333333333
AA:AA:AA:AA:AA:AA (unknown)
...
```
leinfo output: 
```
Requesting information ...
        Handle: 64 (0x0040)
        LMP Version: 4.2 (0x8) LMP Subversion: 0x8
        Manufacturer: Telink Semiconductor Co. Ltd (529)
        Features: 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00
```

So yeah definitely seems like Telink.

I found this post that seem helpful:

https://www.jaredwolff.com/get-started-with-bluetooth-low-energy/

If you have any ideas on useful commands to run, please let me know. 
Here's the results from some of the stuff in that link. Note I modified the UUIDs a little just in case there's personal info in there.
```
# gatttool -b AA:AA:AA:AA:AA:AA -I
[AA:AA:AA:AA:AA:AA][LE]> connect
Attempting to connect to AA:AA:AA:AA:AA:AA
Connection successful
[AA:AA:AA:AA:AA:AA][LE]> primary
attr handle: 0x0001, end grp handle: 0x0006 uuid: 00001900-0000-1000-8000-00105f9b34fc
attr handle: 0x0007, end grp handle: 0x000f uuid: 0000190b-0000-1000-8000-00105f9b34fc
attr handle: 0x0010, end grp handle: 0x001c uuid: 00010203-0405-0607-0309-0a0b0c0d1920
attr handle: 0x001d, end grp handle: 0x001f uuid: 19200d0c-0b0a-0908-0506-050403020111
```

**Comment by juanboro** [comment]
Ok - excellent --- definitely getting closer - we should be able to get this to work.  We can both do some more reading on a few details :-).  One issue I remember with my plug which has both the wifi and ble is that it had the macs backwards at one point (my script grabbed the wifi one incorrectly) - but it looks like it got the right one above.

**Comment by sarnex** [comment]
Yeah BT mac definitely seems right. I'll try to go over the current implementation and see if I can spot anything, but it's very unlikely I'll find anything because I have no idea what I'm doing. You've been very helpful so far, and if you have even more time I'd really appreciate your help,

**Comment by sarnex** [comment]
I just found this repo: https://github.com/vpaeder/telinkpp
And I get the same error using their implementation:

```
terminate called after throwing an instance of 'tinyb::BluetoothException'
  what():  GDBus.Error:org.bluez.Error.Failed: Operation failed with ATT error: 0x0e
Aborted
```

I used AA:AA:AA:AA:AA:AA for the mac, 3333333333333333 for the device name and 222222 for the password, that seems to be right.
```
./telink_test 
Run as: ./telink_test <device_MAC_address> <device_name> <device_password>

```

**Comment by juanboro** [comment]
>I used AA:AA:AA:AA:AA:AA for the mac, 3333333333333333 for the device name and 222222 for the password
Should be -- I'd think you'd get a different error on the pair step if the password is wrong.

**Comment by sarnex** [comment]
You're not going to believe this, but just commenting out
```
await self.client.start_notify(atelink_mesh.notification_char, self.callback_handler)
```

actually fixed it. I had to also comment out the 
```
if not self.online: return False
```
check in set_power because I guess it's marked as offline for some reason that I didn't investigate, but wow.

I think what's happening now is on every set_power event it does a full connection [here](https://github.com/juanboro/cync2mqtt/blob/main/src/acync/mesh.py#L230) so it doesn't need the notify? No idea.

My guess is this link is related:

https://github.com/hbldh/bleak/issues/972#issuecomment-1235867382

TLDR, seems like the hardware doesn't implement the spec correctly, which is sort of what I expected.

I have no idea what effect skipping start_notify has, but at least for now it seems to work.

The device still shows as offline in HA and only mosquitto_pub works, but I can probably fix that with some hacks to the cync2mqtt script.

Thanks so much for your help! If you want to email me (sarnex@gentoo.org) somewhere to donate I can toss you a small token of appreciation.

Closing this as it doesn't seem like a problem in the code.

**Comment by juanboro** [comment]
Awesome!  I'm going to keep it open as an enhancement to see if we can identify the switch type and then make the appropriate adjustments.  I'm still hoping there is a way to get the state notification callbacks.  Let me check that bleak issue out.

I'm just glad some folks can find all this useful like I do :-)  Most of the credit for this all working goes to all the underlying libraries like bleak and amqtt which make all this possible.

**Comment by sarnex** [comment]
Just a final update from my side, with some hacks to the cync2mqtt script to get the device to always report as online to mqtt even if device.online is false, it shows up as online in home assistant as expected and works. I was worried there might be some timeout problem (like it works for the first X minutes after I start the script but then stops working because the start_notify was doing some keep-alive) but that does not seem to be the case, I can wait hours after starting the cync2mqtt script and still control it from home assistant. So everything is working perfectly for me. Thanks again

**Comment by juanboro** [comment]
Great to hear --- yeah - I want to poke into some details as soon as I get some time to see more if there's some way we can fix the notifications for this (also have it properly identify as a plug so that it gets treated as a switch by default in HA).

**Comment by juanboro** [comment]
> Thanks for the response.
> 
> Here's my config, with all info changed obviously.
> 
> ```
> meshconfig:
>   1111111111:
>     access_key: 222222
>     bulbs:
>       1:
>         mac: AA:AA:AA:AA:AA:AA
>         name: Switch
>     mac: 3333333333333333
>     name: why
> mqtt_url: mqtt://username:pass@127.0.0.1:1883/
> ```
> 
> lescan output:
> 
> ```
> ...
> AA:AA:AA:AA:AA:AA 3333333333333333
> AA:AA:AA:AA:AA:AA (unknown)
> ...
> ```
> 
> leinfo output:
> 
> ```
> Requesting information ...
>         Handle: 64 (0x0040)
>         LMP Version: 4.2 (0x8) LMP Subversion: 0x8
>         Manufacturer: Telink Semiconductor Co. Ltd (529)
>         Features: 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00
> ```
> 
> So yeah definitely seems like Telink.
> 
> I found this post that seem helpful:
> 
> https://www.jaredwolff.com/get-started-with-bluetooth-low-energy/
> 
> If you have any ideas on useful commands to run, please let me know. Here's the results from some of the stuff in that link. Note I modified the UUIDs a little just in case there's personal info in there.
> 
> ```
> # gatttool -b AA:AA:AA:AA:AA:AA -I
> [AA:AA:AA:AA:AA:AA][LE]> connect
> Attempting to connect to AA:AA:AA:AA:AA:AA
> Connection successful
> [AA:AA:AA:AA:AA:AA][LE]> primary
> attr handle: 0x0001, end grp handle: 0x0006 uuid: 00001900-0000-1000-8000-00105f9b34fc
> attr handle: 0x0007, end grp handle: 0x000f uuid: 0000190b-0000-1000-8000-00105f9b34fc
> attr handle: 0x0010, end grp handle: 0x001c uuid: 00010203-0405-0607-0309-0a0b0c0d1920
> attr handle: 0x001d, end grp handle: 0x001f uuid: 19200d0c-0b0a-0908-0506-050403020111
> ```

Was taking a look at this again.  What does gatttool return when you type characteristics?

**Comment by sarnex** [comment]
See below, thanks

```
handle: 0x0002, char properties: 0x02, char value handle: 0x0003, uuid: 00002a00-0000-1000-8000-00805f9b34fb
handle: 0x0005, char properties: 0x02, char value handle: 0x0006, uuid: 00002a01-0000-1000-8000-00805f9b34fb
handle: 0x0008, char properties: 0x02, char value handle: 0x0009, uuid: 00002a26-0000-1000-8000-00805f9b34fb
handle: 0x000a, char properties: 0x02, char value handle: 0x000b, uuid: 00002a29-0000-1000-8000-00805f9b34fb
handle: 0x000c, char properties: 0x02, char value handle: 0x000d, uuid: 00002a24-0000-1000-8000-00805f9b34fb
handle: 0x000e, char properties: 0x02, char value handle: 0x000f, uuid: 00002a27-0000-1000-8000-00805f9b34fb
handle: 0x0011, char properties: 0x1a, char value handle: 0x0012, uuid: 00010203-0405-0607-0809-0a0b0c0d1911
handle: 0x0014, char properties: 0x0e, char value handle: 0x0015, uuid: 00010203-0405-0607-0809-0a0b0c0d1912
handle: 0x0017, char properties: 0x06, char value handle: 0x0018, uuid: 00010203-0405-0607-0809-0a0b0c0d1913
handle: 0x001a, char properties: 0x0a, char value handle: 0x001b, uuid: 00010203-0405-0607-0809-0a0b0c0d1914
handle: 0x001e, char properties: 0x02, char value handle: 0x001f, uuid: 00002af0-0000-1000-8000-00805f9b34fb
```

**Comment by juanboro** [comment]
Update: I'm seeing pretty much the exact same issue with notifications when I connect to the direct connect bulb I have.  Now just need to figure out why it doesn't work (bluez issue?)

It's not clear if this has the overlapping handle problem at https://github.com/hbldh/bleak/issues/972

![image](https://github.com/juanboro/cync2mqtt/assets/29214674/b9a2299f-6dff-4666-946d-89916ca457ea)

5-AUG - edit: actually, I think it is the same problem as the issue linked above.  UUID 00010203-0405-0607-0809-0a0b0c0d1911 shows up under multiple service uuids with and without notify (we want service id 00010203-0405-0607-0809-0a0b0c0d1910).  That seems like a hardware bug... but the question is whether this can be worked around with bluez or whether bluez can be bypassed.  I'm investigating several possibilities.

**Comment by juanboro** [comment]
I think this should now work by using bluepy instead of bleak for bluetooth.  See example in: https://github.com/juanboro/cync2mqtt/blob/main/cync_mesh_example.yaml for how to enable this.  Open this issue if it still isn't working.

**Comment by sarnex** [comment]
Yep, it's working with bluepy, thanks!
